SaraAlert/SaraAlert
# Consume job tests

# Description*
Jira Ticket: SARAALERT-586

Adds test for the assessments consumer.

# Important Changes*

* Add tests for the assessments consumer that test that histories and assessments of appropriate types for the given input are being created.
* Add new gem dependency `fakeredis` which mocks the `subscribe` method we are using (PubSub Redis pattern) and allows easier testing.
* Fixes bug with `negate` methods on `IntegerSymptom` and `FloatSymptom` to ensure that they are opposite what the threshold operator is.
* Create new initializer for Redis which allows the Redis object to be shared across the Rails application and provides for configuration later on in the initializer if we desire.

<img width="1601" alt="Screen Shot 2020-08-13 at 09 40 57" src="https://user-images.githubusercontent.com/3009651/90141450-21f0c600-dd49-11ea-9834-b4d6bb7c8943.png">

In the screenshot above you can see me fake the different types of reports that would be coming in from a patient by just submitting the correct values as an epi.

Row 1: Weblink or email reported in detail / does not meet the threshold
Row 2: SMS or voice reported symptomatic / meets the threshold because patient was not specific
Row 3: Weblink or email reported in detail / meets the threshold plus other symptoms flags the entire row
Row 4: SMS or voice reported not experiencing symptoms / does not meet the threshold therefore asymptomatic for this symptom
Row 5: Weblink or email reported in detail / meets threshold.

# Testing*
This fix was tested on the following browsers (submitter must check all those that apply):

Please describe the tests needed to verify your changes. Provide instructions for reproducing these tests. List any relevant details for your test configuration.

```
# To verify the new unit tests
bundle install
rails test
```
To check on the new report table:
* Update the `jurisdiction.yml` to have a Float or Integer symptom
* Submit reports as an Epi with the different cases (Symptomatic, asymptomatic, special numbers, etc.)
* Alternatively you run in split architecture mode and then submit text or web assessments to see the effects on the reports table.

# Submitter Checklist*
- [x] I have tested all of my changes locally. 

# Reviewer Checklist*
## Reviewer 1
Reviewer: holmesie

Checklist:
 - [ ] I have carried out all tests described in the PR and verified functionality. 
 - [ ] I have reviewed every file in the diff to verify that changes are reasonable, and that any updated/added code is maintainable and reusable and follows good practices.

## Reviewer 2
Reviewer: mayerm94

Checklist:
 - [ ] I have carried out all tests described in the PR and verified functionality. 
 - [ ] I have reviewed every file in the diff to verify that changes are reasonable, and that any updated/added code is maintainable and reusable and follows good practices.
